Add UncapturedHigherOrderOpError to always raise exceptions for cond.

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108028
* __->__ #108027
* #108026
* #108025

We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

ydwu4 committed:
Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception

[ghstack-poisoned]

ydwu4 committed:
Update on "Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1037,25 +1037,13 @@ def false_fn(x):
-    gt = getitem > 4;  getitem = None
-    return (gt,)""",
-        )
+        with self.assertRaisesRegex(
```
Btw, does this work with surpress_errors = True?

ydwu4 replied:
yes, GRAPH_BREAK_IN_CONTROL_FLOW check precedes the [suppress_error check](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/convert_frame.py#L658)

zou3519 replied:
Can you add a test that turns suppress_errors=True and check this?

zou3519 commented on torch/_dynamo/exc.py:
```diff
@@ -115,6 +115,7 @@ def __init__(self, msg):
 
 class UserErrorType(Enum):
     DYNAMIC_CONTROL_FLOW = auto()
+    GRAPH_BREAK_IN_CONTROL_FLOW = auto()
```
![image](https://github.com/pytorch/pytorch/assets/5652049/85319318-3f7a-4761-9135-5a1f033427c9)

This shouldn't be a UserErrorType, because it's not supported in eager mode either. The recommendation here is to add a new type of exception. I'm not sure it should subclass Unsupported, it should probably just subclass RuntimeError.

Naming: UncapturableHigherOrderOpError ?

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -313,14 +313,14 @@ def call_function(
         if len(args) != 4:
             raise UserError(
-                UserErrorType.DYNAMIC_CONTROL_FLOW,
+                UserErrorType.GRAPH_BREAK_IN_CONTROL_FLOW,
```
I expected us to have a giant try-catch around the call_function that could raise UncapturableHigherOrderOpError, or some sort of decorator to make it nicer. The error message should say something like "cond doesn't work unless it is captured with torch.compile" and we should rethrow the original error

zou3519 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -654,6 +656,14 @@ def _convert_frame(
             # someone wanted these to also get suppressed.  If so, you'll
             # need to make these exceptions not get wrapped
+
+            # Directly raise if a control flow op (e.g. cond) has a graph break in it.
```
explicitly mention that we don't want to suppress here

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1127,16 +1115,17 @@ def false_fn(x):
 
+        mod_for_eager = Foo()
+        mod_for_eager(torch.tensor(True), torch.tensor(5))
+
```
Btw, we shouldn't have any users of cond out there, but this PR does technically break BC. You might need to be careful when landing (if there are any internal users that are broken)

ydwu4 committed:
Update on "Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception"


We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception"


We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception"


We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

zou3519 reviewed (approved):
LGTM, but change PR title please

ydwu4 committed:
Update on "Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception"


We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Make GRAPH_BREAK_IN_CONTROL_FLOW directly raise exception"


We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Add UncapturedHigherOrderOpError to always raise exceptions for cond."


We want cond to always throw errors despite user's torch.compile mode.

The current implementation is to 
1. catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. 
2. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.
3. At the top-level, we treat it like an hard error and not supressing the errors.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]
